sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala

```diff
@@ -289,6 +289,14 @@ private[hive] class SparkExecuteStatementOperation(
       sqlContext.sparkContext.cancelJobGroup(statementId)
     }
   }
+
+  private def getResultIterator(): Iterator[SparkRow] = {
+    val (totalRowCount, iterResult) = result.collectCountAndIterator()
+    val batchCollectLimit =
+      sqlContext.getConf(SQLConf.THRIFTSERVER_BATCH_COLLECTION_LIMIT.key).toLong
+    resultList = if (totalRowCount < batchCollectLimit) Some(iterResult.toArray) else None
+    if (resultList.isDefined) resultList.get.iterator else iterResult
```

Comment:
When incremental collect is disabled, and users want to use `FETCH_FIRST`, we expect the returned rows are cached and can get iterators again. Now `FETCH_FIRST` will trigger re-execution no matter incremental collect or not. I think this maybe performance regression in some cases.

Reply:
Yes, the case you commented (incremental collect is disabled & FETCH_FIRST) has performance degradation. If total rows are bigger than batchCollectLimit, I thought it is not suitable case for caching decompressed rows because of memory pressure. If FETCH_FIRST caches compressed rows(not decompressed rows) regardless of row count, the result that exceed batchLimit can be cached too. But "Iterator" return type may not a good choice for that. Instead "View" of scala is proper choice, because "Iterator" can be created again with compressed rows "View", but it causes much more source change so I didn't do that.

Reply:
@viirya 
I share my idea of the problem you commented.

1. Change the return type of "collectCountAndIterator" to tuple of (Long, SeqView)
2. The SeqView is created from encoded result array(which is the result of getByteArrayRdd().collect() in SparkPlan), and holds deserializing operations defined in DataSet.
3. Change type of resultList in SparkExecuteStatementOperation to Option[Iterable[SparkRow]], because both Array & SeqView are Iterable.
4. ThriftServer checks if row count exceeds THRIFTSERVER_BATCH_COLLECTION_LIMIT, and decide.
   -> if row count > THRIFTSERVER_BATCH_COLLECTION_LIMIT => resultList cache SeqView.
   -> else resultList caches Array which is collected from SeqView. => resultList cache Array.

How do you think about this idea?

Reply:
I think we should try to cache encoded result if row count >`THRIFTSERVER_BATCH_COLLECTION_LIMIT` when incremental collect is disabled. It sounds to me more close to what the mode does.

Otherwise its behavior looks close as incremental collect mode as it does re-execution. Besides, it collects all data back to driver in encoded format.

Reply:
I will try to cache it. Thank you for reply.